TRIB-196: Updates controller endpoints to return ResponseEntity objects

mrsbluerose opened the pull request:
Related to frontend Jira-TRIB-229;

**Creates a GenericResponseDTO with three properties:**
- responseBoolean
- responseMessage
- responseLong

**FileUploadController:**

- fileIsFound endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseLong to the timestamp
  - Testing: when running program, test log printed on front end, and picture displays
- deleteFile endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseBoolean
  - Testing: cannot find an instance where this endpoint is used
- handleFileUpload endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseMessage
  - Retools error handling to include return statements for possible 400 and 500 errors
  - Testing: set the application.properties picture directory property to a local file and successfully uploaded and changed photos with the expected log messages, including an error message when the resource type was intentionally set incorrectly.

**Profile Controller:**

- update endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseBoolean
  - Testing: ran front end and successfully updated username, picture and phone number with expected log messages

**SMSChallenge Controller**

- sendSMSChallengeCode endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseMessage
  - Testing: successfully changed phone number and created new user, receiving expected log messages and prompts
- clearSMSChallengeCode endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseMessage
  - Testing: did not find a corresponding call on the front end
- isAValidSMSChallengeCode
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseBoolean
  - Testing: successfully changed phone number and created new user with expected log messages

**User Controller**

- Creates UserData enum for use with helper method 
- Creates helper method isUserDataAvailable which checks for exiting user data duplication 
- Deletes individual endpoints isUsernameAvailable, isPhoneNumberAvailable, isEmailAddressAvailable in favor of using new helper method

- isUserInformationUnique endpoint
  - Changes return type to ResponseEntity<GenericResponseDTO>
  - Sets responseMessage
  - Retools method to use helper method isUserDataAvailable
  - Testing: successfully created new user on front end and received expected prompts when purposefully entering duplicate name, email and phone.

**SMS Text Messaging Service**
- Added log message to include a “pretend” message for testing SMS validation codes.
- Testing: Successfully able to enter expected sms code

mrsbluerose committed:
J-TRIB-196: adds GenericMessageDTO

mrsbluerose committed:
J-TRIB-196: updates file cntrl upload endpoint
- updates handleFileUpload endpoint to return GenericMessasgeDTO

mrsbluerose committed:
J-TRIB-196: updates database driver

mrsbluerose committed:
J-TRIB-196: updates FileUploadController upload
- updates handleFileUpload endpoint return DTO string format

mrsbluerose committed:
J-TRIB-196: updates GenericMessageDTO name
- updates GenericMessageDTO to GenericResponseDTO
- updates all references and variables using GenericResponseDTO

mrsbluerose committed:
J-TRIB-196: adds boolean to GenericResponseDTO

mrsbluerose committed:
J-TRIB-196: updates ProfileCntr update endpoint
- changes update endpoint return type from boolean to GenericResponseDTO

mrsbluerose committed:
J-TRIB-196: updates generic DTO declaration
- optimizes setting the responseBoolean property

mrsbluerose committed:
J-TRIB-196: updates SMS Cntrl send code endpoint
- updates SMSChallengeCodeAPIController sendSMSChallengeCode to return a response entity and generic response DTO

mrsbluerose committed:
Merge branch 'main' into Jira-TRIB-196

mrsbluerose committed:
J-TRIB-196: updates SMS Cntrl clear endpoint
- updates SMSChallengeCodeAPIController clearSMSChallengeCode endpoint to return ResponseEntity<GenericResponseDTO>

mrsbluerose committed:
J-TRIB-196: updates SMS conroller isValid endpoint
- updates SMSChallengeCodeAPIController isValidSMSChallengeCode to return ResponseEntity<GenericResponseDTO>

mrsbluerose committed:
TRIB-196: WIP refactor user properties checks

mrsbluerose committed:
J-TRIB-196: adds logger sms challenge code

mrsbluerose committed:
J-TRIB-196: capitalizes user enum property

mrsbluerose committed:
J-TRIB-196: updates userAPI isUnique endpoint
- updates isUserInformationUnique to use a helper method IsUserDataAvailable and enums instead of using other endpoints to check for each user property
- updates helper method to check for isEmpty() with each call to the user repo
- reasoning: The goal is to have each endpoint return a response entity object, and it made more sense to create a helper method rather than having one endpoint call other endpoints.

mrsbluerose committed:
Merge branch 'J-TRIB-UserAPI' into Jira-TRIB-196

mrsbluerose committed:
Revert "J-TRIB-196: updates database driver"

This reverts commit 7c649f67757dcd91aa467327fab82927341de936.
- created separate ticket/pull request for this specific change

mrsbluerose committed:
Merge branch 'main' into Jira-TRIB-196

mrsbluerose committed:
J-TRIB-196: fixed user cntrl helper method
- updated isUserDataAvailable method check for phone number to recognize a list instead of a single value returned from searching the User repo

mrsbluerose committed:
J-TRIB-196: adds long property to Generic DTO

mrsbluerose committed:
J-TRIB-196: updates file cntrl found return type
- updates FileUploadController FileIsFound endpoint to return a response entity with GenericDTO long

mrsbluerose committed:
J-TRIB-196: updates File cntrl delete return type
- updates FileUploadController deleteFile endpoint to return Response Entity with GenericResponseDTO boolean

mrsbluerose committed:
J-TRIB-196: updates file cntrl upload rtn type
- updates FileUploadController handleFileUpload endpoint to return response entity with GenericResponseDTO

mrsbluerose committed:
J-TRIB-196: updates handlFileUpload error handling
- creates stand alone if statement to verify that the resource type is correct, and sends back a 400 bad request if not
- calls storage and picture services in try block and sends back a 200 okay
- catches IO exception and sends 500 internal server error if not okay
- reasoning: in Jan 4 2024's meeting we discussed handling bad request and server errors in this endpoint. I researched errors and consulted ChatGPT. It recommended removing my redundant finally block and handling the 400 error first and foremost before entering the try/catch.

mrsbluerose committed:
J-TRIB-196: removes non-public user data endpoints
- removes endpoints that checked for existing user name, email and phone
- reasoning: The isUserInformationUnique endpoint made calls to individual endpoints to check for each data type. An existing TODO note wanted to find a way to use these methods without granting public access. Rather than using endpoints (and handling the response entity object return types), I created a private helper method that checks for each user data type.

mrsbluerose committed:
J-TRIB-196: updates DTO builder for consistency
- updates the update endpoint DTO builder to have each part of the builder on seperate lines

mrsbluerose committed:
J-TRIB-196: updates sms send DTO builder
- updates DTO builder to have each dot method on a seperate line

mrsbluerose committed:
J-TRIB-196: updates sms clear DTO builder
- updates DTO builder to have each dot method on a seperate line

mrsbluerose committed:
J-TRIB-196: updates sms isValid DTO builder
- updates DTO builder to have each dot method on a seperate line

haxwell commented on src/main/java/com/savvato/basemobileapp/controllers/ProfileAPIController.java:
```diff
@@ -33,13 +34,17 @@ public ResponseEntity<ProfileDTO> getById(@PathVariable Long profileId) {
+
+		GenericResponseDTO genericResponseDTO = GenericResponseDTO
+				.builder()
+				.responseBoolean(profileService.update(request.userId, request.name, request.email, request.phone))
```
Use a variable here. It will make debugging easier in the future. Also, where you created GenericResponseDTOs above here, you used a variable. Not a big deal.

mrsbluerose replied:
I'll do the variable. Thanks for pointing this out. I was conflicted, and I like the other instances where I created a variable and then used it in the DTO builder.

haxwell commented on src/main/java/com/savvato/basemobileapp/controllers/SMSChallengeCodeAPIController.java:
```diff
@@ -21,32 +24,42 @@ public class SMSChallengeCodeAPIController {
-        return rtn;
+        GenericResponseDTO genericResponseDTO = GenericResponseDTO
+                .builder()
+                .responseMessage(smsccs.sendSMSChallengeCodeToPhoneNumber(phoneNumber))
```
Here as well, use a variable.

haxwell replied:
Or alternatively, you could inline the other places where you did use a variable. Not as helpful for the next developer, but not a bad choice either. Consistency is the key.

haxwell commented on src/main/java/com/savvato/basemobileapp/controllers/SMSChallengeCodeAPIController.java:
```diff
@@ -21,32 +24,42 @@ public class SMSChallengeCodeAPIController {
-        return smsccs.isAValidSMSChallengeCode(req.phoneNumber, req.code);
+        GenericResponseDTO genericResponseDTO = GenericResponseDTO
+                .builder()
+                .responseBoolean(smsccs.isAValidSMSChallengeCode(req.phoneNumber, req.code))
```
.

haxwell commented on src/main/java/com/savvato/basemobileapp/controllers/UserAPIController.java:
```diff
@@ -48,43 +51,31 @@ public ResponseEntity createUser(@RequestBody @Valid UserRequest req) {
+	}
+
+	private boolean isUserDataAvailable(UserData typeOfData, String data){
+		boolean rtn = switch(typeOfData) {
```
This is a very pretty refactor. Well done.

mrsbluerose replied:
Thanks! I'm trying to get more comfortable with the new switch statement, and this seemed like a good use case. (As long as I remember how to handle returned optionals haha)

haxwell commented:
Overall, this is a great PR. Some minor formatting-type issues.

mrsbluerose committed:
J-TRIB-196: creates profile cntrl update variable
- sets ProfileAPIController update endpoint call to the SMSChallengeCodeService to a variable and then uses variable in the DTO builder

mrsbluerose committed:
J-TRIB-196: creates SMS cntrl isValid variable
- sets SMSChallengeCodeAPIController isAValidSMSChallengeCoded endpoint call to the SMSChallengeCodeService to a variable and then uses variable in the DTO builder

mrsbluerose committed:
J-TRIB-196: creates SMS cntrl send code variable
- sets SMSChallengeCodeAPIController sendSMSChallengeCode endpoint call to the SMSChallengeCodeService to a variable and then uses variable in the DTO builder

haxwell commented on src/main/java/com/savvato/basemobileapp/controllers/ProfileAPIController.java:
```diff
@@ -33,13 +34,17 @@ public ResponseEntity<ProfileDTO> getById(@PathVariable Long profileId) {
-		boolean rtn = profileService.update(request.userId, request.name, request.email, request.phone);
+	public ResponseEntity<GenericResponseDTO> update(@RequestBody @Valid ProfileRequest request) {
+		Boolean isUpdated = profileService.update(request.userId, request.name, request.email, request.phone);
+		GenericResponseDTO genericResponseDTO = GenericResponseDTO
```
This works, but it makes more sense to build the dto after checking if the return value was as expected.
